sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ApproximateQuantilesTest.java

```diff
@@ -35,283 +37,385 @@
   }
 
-  private static class OrderByLength implements Comparator<String>, Serializable {
-    @Override
-    public int compare(String a, String b) {
-      if (a.length() != b.length()) {
-        return a.length() - b.length();
-      } else {
-        return a.compareTo(b);
+  /** Tests to ensure we are calculating the optimal buffers. */
+  @RunWith(Parameterized.class)
+  public static class BufferTests {
+    private static final Logger LOG = LoggerFactory.getLogger(BufferTests.class);
+
+    private final double epsilon;
+    private final long maxInputSize;
+    private final int expectedNumBuffers;
+    private final int expectedBufferSize;
+    private final ApproximateQuantilesCombineFn<?, ?> combineFn;
+
+    /**
+     * Test data taken from "Munro-Paterson Algorithm" reference values table of "Approximate
+     * Medians and other Quantiles in One Pass and with Limited Memory" paper.
+     *
+     * @see ApproximateQuantilesCombineFn for paper reference.
+     */
+    private static final double[] epsilons = new double[] { 0.1, 0.05, 0.01, 0.005, 0.001 };
+    private static final int[] maxElementExponents = new int[] { 5, 6, 7, 8, 9 };
+
+    private static final int[][] expectedNumBuffersValues = new int[][] {
+        { 11, 14, 17, 21, 24 },
+        { 11, 14, 17, 20, 23 },
+        {  9, 11, 14, 17, 21 },
+        {  8, 11, 14, 17, 20 },
+        {  6,  9, 11, 14, 17},
+    };
+
+    private static final int[][] expectedBufferSizeValues = new int[][] {
+        {   98,  123,  153,    96,   120 },
+        {   98,  123,  153,   191,   239 },
+        {  391,  977, 1221,  1526,   954 },
+        {  782,  977, 1221,  1526,  1908 },
+        { 3125, 3907, 9766, 12208, 15259 },
+    };
+
+    @Parameterized.Parameters
+    public static Collection<Object[]> data() {
+      Collection<Object[]> testData = Lists.newArrayList();
+      for (int i = 0; i < epsilons.length; i++) {
+        for (int j = 0; j < maxElementExponents.length; j++) {
+          testData.add(new Object[]{
+              epsilons[i],
+              (long) Math.pow(10, maxElementExponents[j]),
+              expectedNumBuffersValues[i][j],
+              expectedBufferSizeValues[i][j]
+          });
+        }
       }
+
+      return testData;
     }
-  }
 
+    public BufferTests(
+        Double epsilon, Long maxInputSize, Integer expectedNumBuffers, Integer expectedBufferSize) {
+      this.epsilon = epsilon;
+      this.maxInputSize = maxInputSize;
+      this.expectedNumBuffers = expectedNumBuffers;
+      this.expectedBufferSize = expectedBufferSize;
 
-  private PCollection<Integer> intRangeCollection(Pipeline p, int size) {
-    return p.apply("CreateIntsUpTo(" + size + ")", Create.of(intRange(size)));
-  }
+      this.combineFn = ApproximateQuantilesCombineFn.create(
+          10, new Top.Largest<Long>(), maxInputSize, epsilon);
+    }
+
+    @Before
+    public void setUp() {
+      LOG.info("Epsilon: {}, Max input size: {}", epsilon, maxInputSize);
```

Comment:
please drop. Instead, you can do this directly in the `@Parameters` annotation, then you also see which parameters failed on the command-line (or in your IDE) when it does.

Example: https://github.com/apache/incubator-beam/blob/064796dcc6bffe5028411006c196129ee8766c44/sdks/java/core/src/test/java/org/apache/beam/sdk/io/range/ByteKeyRangeEstimateFractionTest.java#L38)